// @avnu/avnu-sdk/dist/index.mjs
// src/constants.ts
var SEPOLIA_BASE_URL = "https://sepolia.api.avnu.fi";
var BASE_URL = "https://starknet.api.avnu.fi";

// src/fixtures.ts
import { parseUnits } from "ethers";
import { constants } from "starknet";

// src/types.ts
var SourceType = /* @__PURE__ */ ((SourceType2) => {
  SourceType2["DEX"] = "DEX";
  SourceType2["MARKET_MAKER"] = "MARKET_MAKER";
  SourceType2["SOLVER"] = "SOLVER";
  return SourceType2;
})(SourceType || {});
var ContractError = class {
  constructor(message, revertError) {
    this.message = message;
    this.revertError = revertError;
  }
};

// src/fixtures.ts
var aPriceRequest = () => ({
  sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
  sellAmount: parseUnits("1", 18),
  buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7"
});
var aQuoteRequest = () => ({
  sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
  sellAmount: parseUnits("1", 18),
  buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
  size: 1,
  takerAddress: "0x0"
});
var aPrice = () => ({
  sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
  sellAmount: parseUnits("1", 18),
  sellAmountInUsd: 1700,
  buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
  buyAmount: parseUnits("2", 18),
  buyAmountInUsd: 1700,
  blockNumber: 1,
  chainId: constants.StarknetChainId.SN_SEPOLIA,
  sourceName: "AMM1",
  priceRatioUsd: 0,
  gasFees: BigInt(0),
  gasFeesInUsd: 0
});
var aQuote = () => ({
  quoteId: "quoteId",
  sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
  sellAmount: parseUnits("1", 18),
  sellAmountInUsd: 1700,
  buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
  buyAmount: parseUnits("2", 18),
  buyAmountInUsd: 1700,
  buyAmountWithoutFees: parseUnits("2", 18),
  buyAmountWithoutFeesInUsd: 1700,
  blockNumber: 1,
  chainId: constants.StarknetChainId.SN_SEPOLIA,
  expiry: 1e11,
  routes: [
    {
      name: "AMM1",
      address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c661",
      percent: 1,
      sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
      buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
      routes: []
    }
  ],
  gasFees: BigInt("0x0"),
  gasFeesInUsd: 0,
  avnuFees: BigInt("0x0"),
  avnuFeesInUsd: 0,
  avnuFeesBps: BigInt("0x0"),
  integratorFees: BigInt("0x0"),
  integratorFeesInUsd: 0,
  integratorFeesBps: BigInt("0x0"),
  priceRatioUsd: 0,
  liquiditySource: "DEX_AGGREGATOR",
  gasless: {
    active: false,
    gasTokenPrices: []
  }
});
var aQuoteWithManySubRoutes = () => ({
  quoteId: "quoteId",
  sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
  sellAmount: parseUnits("1", 18),
  sellAmountInUsd: 1700,
  buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
  buyAmount: parseUnits("2", 18),
  buyAmountInUsd: 1700,
  buyAmountWithoutFees: parseUnits("2", 18),
  buyAmountWithoutFeesInUsd: 1700,
  blockNumber: 1,
  chainId: constants.StarknetChainId.SN_SEPOLIA,
  expiry: 1e11,
  priceRatioUsd: 0,
  liquiditySource: "DEX_AGGREGATOR",
  routes: [
    {
      name: "AMM1",
      address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c661",
      percent: 1,
      sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
      buyTokenAddress: "0x3e85bfbb8e2a42b7bead9e88e9a1b19dbccf661471061807292120462396ec9",
      routes: [
        {
          name: "AMM2",
          address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c662",
          percent: 1,
          sellTokenAddress: "0x3e85bfbb8e2a42b7bead9e88e9a1b19dbccf661471061807292120462396ec9",
          buyTokenAddress: "0x2e2faab2cad8ecdde5e991798673ddcc08983b872304a66e5f99fbb24e14abc",
          routes: [
            {
              name: "AMM1",
              address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c661",
              percent: 1,
              sellTokenAddress: "0x2e2faab2cad8ecdde5e991798673ddcc08983b872304a66e5f99fbb24e14abc",
              buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
              routes: []
            }
          ]
        }
      ]
    }
  ],
  gasFees: BigInt("0x0"),
  gasFeesInUsd: 0,
  avnuFees: BigInt("0x0"),
  avnuFeesInUsd: 0,
  avnuFeesBps: BigInt("0x0"),
  integratorFees: BigInt("0x0"),
  integratorFeesInUsd: 0,
  integratorFeesBps: BigInt("0x0"),
  gasless: {
    active: false,
    gasTokenPrices: []
  }
});
var aQuoteWithManyComplexRoutes = () => ({
  quoteId: "quoteId",
  sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
  sellAmount: parseUnits("1", 18),
  sellAmountInUsd: 1700,
  buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
  buyAmount: parseUnits("2", 18),
  buyAmountInUsd: 1700,
  buyAmountWithoutFees: parseUnits("2", 18),
  buyAmountWithoutFeesInUsd: 1700,
  blockNumber: 1,
  chainId: constants.StarknetChainId.SN_SEPOLIA,
  expiry: 1e11,
  gasFees: BigInt("0x0"),
  gasFeesInUsd: 0,
  avnuFees: BigInt("0x0"),
  avnuFeesInUsd: 0,
  avnuFeesBps: BigInt("0x0"),
  integratorFees: BigInt("0x0"),
  integratorFeesInUsd: 0,
  integratorFeesBps: BigInt("0x0"),
  priceRatioUsd: 0,
  liquiditySource: "DEX_AGGREGATOR",
  routes: [
    {
      name: "AMM1",
      address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c661",
      percent: 0.5,
      sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
      buyTokenAddress: "0x3e85bfbb8e2a42b7bead9e88e9a1b19dbccf661471061807292120462396ec9",
      routes: [
        {
          name: "AMM2",
          address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c662",
          percent: 0.5,
          sellTokenAddress: "0x3e85bfbb8e2a42b7bead9e88e9a1b19dbccf661471061807292120462396ec9",
          buyTokenAddress: "0x2e2faab2cad8ecdde5e991798673ddcc08983b872304a66e5f99fbb24e14abc",
          routes: [
            {
              name: "AMM1",
              address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c661",
              percent: 1,
              sellTokenAddress: "0x2e2faab2cad8ecdde5e991798673ddcc08983b872304a66e5f99fbb24e14abc",
              buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
              routes: []
            }
          ]
        },
        {
          name: "AMM1",
          address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c661",
          percent: 0.5,
          sellTokenAddress: "0x3e85bfbb8e2a42b7bead9e88e9a1b19dbccf661471061807292120462396ec9",
          buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
          routes: []
        }
      ]
    },
    {
      name: "AMM1",
      address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c661",
      percent: 0.2,
      sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
      buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
      routes: []
    },
    {
      name: "AMM1",
      address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c661",
      percent: 0.3,
      sellTokenAddress: "0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
      buyTokenAddress: "0x3e85bfbb8e2a42b7bead9e88e9a1b19dbccf661471061807292120462396ec9",
      routes: [
        {
          name: "AMM2",
          address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c662",
          percent: 0.2,
          sellTokenAddress: "0x3e85bfbb8e2a42b7bead9e88e9a1b19dbccf661471061807292120462396ec9",
          buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
          routes: []
        },
        {
          name: "AMM1",
          address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c661",
          percent: 0.8,
          sellTokenAddress: "0x3e85bfbb8e2a42b7bead9e88e9a1b19dbccf661471061807292120462396ec9",
          buyTokenAddress: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
          routes: []
        }
      ]
    }
  ],
  gasless: {
    active: false,
    gasTokenPrices: []
  }
});
var anInvokeSwapResponse = () => ({
  transactionHash: "0x0"
});
var aBuildSwapTransaction = () => ({
  chainId: constants.StarknetChainId.SN_SEPOLIA,
  calls: [
    {
      contractAddress: "0x0",
      entrypoint: "execute",
      calldata: []
    }
  ]
});
var ethToken = () => ({
  name: "Ethereum",
  address: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
  symbol: "ETH",
  decimals: 18,
  chainId: "0x534e5f474f45524c49",
  logoUri: "https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png"
});
var btcToken = () => ({
  name: "Wrapped Bitcoin",
  address: "0x72df4dc5b6c4df72e4288857317caf2ce9da166ab8719ab8306516a2fddfff7",
  symbol: "WBTC",
  decimals: 18,
  chainId: "0x534e5f474f45524c49",
  logoUri: "https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599/logo.png"
});
var aPage = (content, size = 10, number = 0, totalPages = 1, totalElements = 1) => ({
  content,
  size,
  totalPages,
  number,
  totalElements
});
var aSource = () => ({
  name: "AMM1",
  address: "0x975910cd99bc56bd289eaaa5cee6cd557f0ddafdb2ce6ebea15b158eb2c661",
  icon: "https://pbs.twimg.com/profile_images/1567441002063069184/SGtDtW-C_400x400.jpg",
  type: "DEX" /* DEX */
});

// src/services.ts
import { toBeHex } from "ethers";
import qs from "qs";
import { ec, hash } from "starknet";
var getBaseUrl = () => process.env.NODE_ENV === "dev" ? SEPOLIA_BASE_URL : BASE_URL;
var parseResponse = (response, avnuPublicKey) => {
  if (response.status === 400) {
    return response.json().then((error) => {
      throw new Error(error.messages[0]);
    });
  }
  if (response.status === 500) {
    return response.json().then((error) => {
      if (error.messages.length >= 0 && error.messages[0].includes("Contract error")) {
        throw new ContractError(error.messages[0], error.revertError || "");
      } else {
        throw new Error(error.messages[0]);
      }
    });
  }
  if (response.status > 400) {
    throw new Error(`${response.status} ${response.statusText}`);
  }
  if (avnuPublicKey) {
    const signature = response.headers.get("signature");
    if (!signature)
      throw new Error("No server signature");
    return response.clone().text().then((textResponse) => {
      const hashResponse = hash.computeHashOnElements([hash.starknetKeccak(textResponse)]);
      const formattedSig = signature.split(",").map((s) => BigInt(s));
      const signatureType = new ec.starkCurve.Signature(formattedSig[0], formattedSig[1]);
      if (!ec.starkCurve.verify(signatureType, hashResponse, avnuPublicKey))
        throw new Error("Invalid server signature");
    }).then(() => response.json());
  }
  return response.json();
};
var fetchPrices = (request, options) => {
  const queryParams = qs.stringify({ ...request, sellAmount: toBeHex(request.sellAmount) }, { arrayFormat: "repeat" });
  return fetch(`${options?.baseUrl ?? getBaseUrl()}/swap/v2/prices?${queryParams}`, {
    signal: options?.abortSignal,
    headers: { ...options?.avnuPublicKey !== void 0 && { "ask-signature": "true" } }
  }).then((response) => parseResponse(response, options?.avnuPublicKey)).then(
    (prices) => prices.map((price) => ({
      ...price,
      sellAmount: BigInt(price.sellAmount),
      buyAmount: BigInt(price.buyAmount),
      gasFees: BigInt(price.gasFees)
    }))
  );
};
var fetchQuotes = (request, options) => {
  if (!request.sellAmount && !request.buyAmount)
    throw new Error("Sell amount or buy amount is required");
  const queryParams = qs.stringify(
    {
      ...request,
      buyAmount: request.buyAmount ? toBeHex(request.buyAmount) : void 0,
      sellAmount: request.sellAmount ? toBeHex(request.sellAmount) : void 0,
      integratorFees: request.integratorFees ? toBeHex(request.integratorFees) : void 0
    },
    { arrayFormat: "repeat" }
  );
  return fetch(`${options?.baseUrl ?? getBaseUrl()}/swap/v2/quotes?${queryParams}`, {
    signal: options?.abortSignal,
    headers: { ...options?.avnuPublicKey !== void 0 && { "ask-signature": "true" } }
  }).then((response) => parseResponse(response, options?.avnuPublicKey)).then(
    (quotes) => quotes.map((quote) => ({
      ...quote,
      sellAmount: BigInt(quote.sellAmount),
      buyAmount: BigInt(quote.buyAmount),
      buyAmountWithoutFees: BigInt(quote.buyAmountWithoutFees),
      gasFees: BigInt(quote.gasFees),
      avnuFees: BigInt(quote.avnuFees),
      integratorFees: BigInt(quote.integratorFees),
      avnuFeesBps: BigInt(quote.avnuFeesBps),
      integratorFeesBps: BigInt(quote.integratorFeesBps),
      gasless: quote.gasless && {
        active: quote.gasless.active,
        gasTokenPrices: quote.gasless.gasTokenPrices.map((gasTokenPrice) => ({
          tokenAddress: gasTokenPrice.tokenAddress,
          gasFeesInUsd: gasTokenPrice.gasFeesInUsd,
          gasFeesInGasToken: BigInt(gasTokenPrice.gasFeesInGasToken)
        }))
      }
    }))
  );
};
var fetchExecuteSwapTransaction = (quoteId, signature, options) => {
  if (Array.isArray(signature)) {
    signature = signature.map((sig) => toBeHex(BigInt(sig)));
  } else if (signature.r && signature.s) {
    signature = [toBeHex(BigInt(signature.r)), toBeHex(BigInt(signature.s))];
  }
  return fetch(`${options?.baseUrl ?? getBaseUrl()}/swap/v2/execute`, {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      ...options?.avnuPublicKey && { "ask-signature": "true" }
    },
    body: JSON.stringify({ quoteId, signature })
  }).then((response) => parseResponse(response, options?.avnuPublicKey));
};
var fetchBuildExecuteTransaction = (quoteId, takerAddress, slippage, includeApprove, options) => fetch(`${options?.baseUrl ?? getBaseUrl()}/swap/v2/build`, {
  method: "POST",
  headers: {
    Accept: "application/json",
    "Content-Type": "application/json",
    ...options?.avnuPublicKey && { "ask-signature": "true" }
  },
  body: JSON.stringify({ quoteId, takerAddress, slippage, includeApprove })
}).then((response) => parseResponse(response, options?.avnuPublicKey));
var fetchBuildSwapTypedData = (quoteId, gasTokenAddress, maxGasTokenAmount, includeApprove = true, takerAddress, slippage, options) => fetch(`${options?.baseUrl ?? getBaseUrl()}/swap/v2/build-typed-data`, {
  method: "POST",
  headers: {
    Accept: "application/json",
    "Content-Type": "application/json",
    ...options?.avnuPublicKey && { "ask-signature": "true" }
  },
  body: JSON.stringify({
    quoteId,
    takerAddress,
    slippage,
    includeApprove,
    gasTokenAddress,
    maxGasTokenAmount: toBeHex(maxGasTokenAmount)
  })
}).then((response) => parseResponse(response, options?.avnuPublicKey));
var fetchTokens = (request, options) => fetch(`${options?.baseUrl ?? getBaseUrl()}/swap/v2/tokens?${qs.stringify(request ?? {})}`, {
  signal: options?.abortSignal,
  headers: { ...options?.avnuPublicKey && { "ask-signature": "true" } }
}).then((response) => parseResponse(response, options?.avnuPublicKey));
var fetchSources = (options) => fetch(`${options?.baseUrl ?? getBaseUrl()}/swap/v2/sources`, {
  signal: options?.abortSignal,
  headers: { ...options?.avnuPublicKey && { "ask-signature": "true" } }
}).then((response) => parseResponse(response, options?.avnuPublicKey));
var executeSwap = async (account, quote, {
  executeApprove = true,
  gasless = false,
  gasTokenAddress,
  maxGasTokenAmount,
  slippage = 5e-3,
  executeGaslessTxCallback
} = {}, options) => {
  const chainId = await account.getChainId();
  if (chainId !== quote.chainId) {
    throw Error(`Invalid chainId`);
  }
  if (gasless) {
    if (!gasTokenAddress || !maxGasTokenAmount) {
      throw Error(`Should provide gasTokenAddress and maxGasTokenAmount when gasless is true`);
    }
    const typedData = await fetchBuildSwapTypedData(
      quote.quoteId,
      gasTokenAddress,
      maxGasTokenAmount,
      executeApprove,
      account.address,
      slippage,
      options
    );
    const signature = await account.signMessage(typedData);
    if (executeGaslessTxCallback) {
      executeGaslessTxCallback();
    }
    return fetchExecuteSwapTransaction(quote.quoteId, signature, options).then((value) => ({
      transactionHash: value.transactionHash,
      gasTokenAddress: value.gasTokenAddress,
      gasTokenAmount: BigInt(value.gasTokenAmount)
    }));
  } else {
    return fetchBuildExecuteTransaction(quote.quoteId, account.address, slippage, executeApprove, options).then(({ calls }) => account.execute(calls)).then((value) => ({ transactionHash: value.transaction_hash }));
  }
};
var calculateMinAmount = (amount, slippage) => amount - amount * BigInt(slippage) / BigInt(1e4);
export {
  BASE_URL,
  ContractError,
  SEPOLIA_BASE_URL,
  SourceType,
  aBuildSwapTransaction,
  aPage,
  aPrice,
  aPriceRequest,
  aQuote,
  aQuoteRequest,
  aQuoteWithManyComplexRoutes,
  aQuoteWithManySubRoutes,
  aSource,
  anInvokeSwapResponse,
  btcToken,
  calculateMinAmount,
  ethToken,
  executeSwap,
  fetchBuildExecuteTransaction,
  fetchBuildSwapTypedData,
  fetchExecuteSwapTransaction,
  fetchPrices,
  fetchQuotes,
  fetchSources,
  fetchTokens
};
